ironfish-cli/src/commands/accounts/repair.ts

```diff
@@ -0,0 +1,183 @@
+    await this.repairBalance(account, node.wallet.walletDb, node.chain)
+
+    this.log('Repairing nullifierToNote')
+    await this.repairNullifierToNoteHash(account, node.wallet.walletDb)
+
+    this.log('Repairing sequenceToNoteHash')
+    await this.repairSequenceToNoteHash(account, node.wallet.walletDb)
+  }
+
+  private loadAccount(wallet: Wallet, accountName: string | undefined): Account {
+    if (accountName) {
+      const account = wallet.getAccountByName(accountName)
+      if (account) {
+        return account
+      }
+      throw new Error(`No account with name ${accountName}`)
+    }
+
+    const defaultAccount = wallet.getDefaultAccount()
+    if (defaultAccount) {
+      return defaultAccount
+    }
+
+    throw new Error('Could not find an account to repair.')
+  }
+
+  private async repairBalance(
+    account: Account,
+    walletDb: WalletDB,
+    chain: Blockchain,
+  ): Promise<void> {
+    let unconfirmedBalance = 0n
+
+    let noteUnspentMismatches = 0
+
+    for await (const decryptedNoteValue of account.getNotes()) {
+      const transactionValue = await account.getTransaction(decryptedNoteValue.transactionHash)
+
+      Assert.isNotUndefined(
+        transactionValue,
+        `Account has a note but is missing the transaction that it received the note from. ${RESCAN_MESSAGE}`,
+      )
+
+      if (!decryptedNoteValue.nullifier) {
+        if (transactionValue.sequence) {
+          throw new Error(
+            `Transaction marked as on chain, but note missing nullifier. ${RESCAN_MESSAGE}`,
+          )
+        }
+
+        continue
+      }
+
+      const spent = await chain.nullifiers.contains(decryptedNoteValue.nullifier)
+
+      if (spent && !decryptedNoteValue.spent) {
+        noteUnspentMismatches++
+
+        await walletDb.saveDecryptedNote(account, decryptedNoteValue.hash, {
+          ...decryptedNoteValue,
+          spent,
+        })
+      } else if (
+        !spent &&
+        !chain.verifier.isExpiredSequence(
+          transactionValue.transaction.expirationSequence(),
+          chain.head.sequence,
+        )
+      ) {
+        unconfirmedBalance += decryptedNoteValue.note.value()
+      }
+    }
+
+    this.log(
+      `\tSaving new unconfirmed balance: ${CurrencyUtils.renderIron(unconfirmedBalance, true)}`,
+    )
+    await walletDb.saveUnconfirmedBalance(account, unconfirmedBalance)
+
+    this.log(
+      `\tRepaired ${noteUnspentMismatches} decrypted notes incorrectly marked as unspent`,
+    )
+  }
+
+  private async repairNullifierToNoteHash(account: Account, walletDb: WalletDB): Promise<void> {
+    let missingNotes = 0
+
+    for await (const [[, nullifier], noteHash] of walletDb.nullifierToNoteHash.getAllIter(
+      undefined,
+      account.prefixRange,
+    )) {
+      const decryptedNoteValue = await account.getDecryptedNote(noteHash)
+
+      if (!decryptedNoteValue) {
+        missingNotes++
+
+        await walletDb.deleteNullifier(account, nullifier)
+      }
```

Comment:
Could we also do this if decryptedNoteValue.nullifier is null?

Reply:
Updated 👍